tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11138,25 +11138,79 @@ bool Sema::DiagnoseHLSLMethodCall(const CXXMethodDecl *MD, SourceLocation Loc) {
         Diags.Report(Loc, diag::err_hlsl_wg_nodetrackrwinputsharing_missing);
         return true;
       }
-    } else if (opCode == hlsl::IntrinsicOp::MOP_CalculateLevelOfDetail ||
-               opCode ==
-                   hlsl::IntrinsicOp::MOP_CalculateLevelOfDetailUnclamped) {
-      const auto *shaderModel =
-          hlsl::ShaderModel::GetByName(getLangOpts().HLSLProfile.c_str());
-      if (!shaderModel->IsSM68Plus()) {
+    }
+  }
+  return false;
+}
+
+// Check HLSL member call constraints for used functions.
+void Sema::DiagnoseUsedHLSLMethodCall(const CXXMethodDecl *MD,
+                                      SourceLocation Loc,
+                                      const hlsl::ShaderModel *SM,
+                                      DXIL::ShaderKind EntrySK,
+                                      const FunctionDecl *EntryDecl) {
+  if (MD->hasAttr<HLSLIntrinsicAttr>()) {
+    hlsl::IntrinsicOp opCode =
+        (IntrinsicOp)MD->getAttr<HLSLIntrinsicAttr>()->getOpcode();
+    switch (opCode) {
+    case hlsl::IntrinsicOp::MOP_CalculateLevelOfDetail:
+    case hlsl::IntrinsicOp::MOP_CalculateLevelOfDetailUnclamped: {
+      if (!SM->IsSM68Plus()) {
         QualType SamplerComparisonTy =
             HLSLExternalSource::FromSema(this)->GetBasicKindType(
                 AR_OBJECT_SAMPLERCOMPARISON);
         if (MD->getParamDecl(0)->getType() == SamplerComparisonTy) {
           Diags.Report(Loc,
-                       diag::err_hlsl_intrinsic_overload_in_wrong_shader_model)
-              << MD->getNameAsString() << "6.8";
-          return true;
+                       diag::warn_hlsl_intrinsic_overload_in_wrong_shader_model)
+              << MD->getNameAsString() + " with SamplerComparisonState"
+              << "6.8";
         }
       }
+      if (!SM->AllowDerivatives(EntrySK)) {
+        Diags.Report(Loc, diag::warn_hlsl_derivatives_in_wrong_shader_kind)
+            << MD->getNameAsString();
+      }
+      switch (EntrySK) {
+      default:
+        break;
+      case DXIL::ShaderKind::Compute:
+      case DXIL::ShaderKind::Amplification:
+      case DXIL::ShaderKind::Mesh: {
+        if (!SM->IsSM66Plus()) {
+          Diags.Report(Loc, diag::warn_hlsl_derivatives_in_wrong_shader_model)
+              << MD->getNameAsString();
```

Comment:
Isn't this unreachable due to the `AllowDerivatives` method checking `IsSM66Plus` already?

Reply:
Moved AllowDerivatives check into the switch.